MaeBlock.cpp

```diff
@@ -44,53 +191,186 @@ bool BufferedIndexedBlockMap::hasIndexedBlock(const std::string& name)
+{
+    size_t count = 0;
+    for(const auto& p : m_bmap) count = max(p.second->size(), count);
+    for(const auto& p : m_imap) count = max(p.second->size(), count);
+    for(const auto& p : m_rmap) count = max(p.second->size(), count);
+    for(const auto& p : m_smap) count = max(p.second->size(), count);
+
+    return count;
+}
+
+
+template <typename T>
+static void output_indexed_property_values(ostream& out,
+        const string& indentation, map<string, T> properties,
+        unsigned int index)
+{
+    for(const auto& p : properties) {
+        const auto& property = p.second;
+        if(property->isDefined(index)) {
+            out << " " << local_to_string(property->at(index));
+        } else {
+            out << " <>";
+        }
+    }
+}
+
+void IndexedBlock::write(ostream& out, unsigned int current_indentation) const
+{
+    string root_indentation = string(current_indentation, ' ');
+    string indentation = string(current_indentation+2, ' ');
+    const bool has_data = m_bmap.size() + m_rmap.size() + m_imap.size() + m_smap.size() > 0;
+
+    out << root_indentation << getName() << "[" << to_string((int)size()) << "] {\n";
+
+    if(has_data) {
+        out << indentation + "# First column is Index #\n";
+    }
+
+    output_property_names(out, indentation, m_bmap);
+    output_property_names(out, indentation, m_rmap);
+    output_property_names(out, indentation, m_imap);
+    output_property_names(out, indentation, m_smap);
+
+    if(has_data) {
+        out << indentation + ":::\n";
+    }
+
+    for(unsigned int i=0; i<size(); i++) {
+        out << indentation << i+1;
+        output_indexed_property_values(out, indentation, m_bmap, i);
+        output_indexed_property_values(out, indentation, m_rmap, i);
+        output_indexed_property_values(out, indentation, m_imap, i);
+        output_indexed_property_values(out, indentation, m_smap, i);
+        out << endl;
+    }
+
+    if(has_data) {
+        out << indentation + ":::\n";
+    }
+
+    out << root_indentation << "}\n";
+
+    return;
+}
+
+string IndexedBlock::toString() const
+{
+    ostringstream stream;
+    write(stream);
+
+    return stream.str();
+}
+
+template <typename T>
+bool IndexedProperty<T>::operator==(const IndexedProperty<T>& rhs) const
+{
+    if(m_is_null == nullptr || rhs.m_is_null == nullptr) {
+        if((m_is_null == nullptr) != (rhs.m_is_null == nullptr)) return false;
+    } else if(*m_is_null != *(rhs.m_is_null)) {
+        return false;
+    }
+    if(m_data != rhs.m_data) return false;
+    return true;
+}
+
+
+// For doubles we need to implement our own comparator for the vectors to
+// take precision into account
+template <>
+bool IndexedProperty<double>::operator==(const IndexedProperty<double>& rhs) const
 {
-    set_indexed_property<IndexedProperty<std::string>>(m_smap, name, value);
+    if(m_is_null == nullptr || rhs.m_is_null == nullptr) {
+        if((m_is_null == nullptr) != (rhs.m_is_null == nullptr)) return false;
```

Comment:
The outer if could be replaced by the inner if:

    if((m_is_null == nullptr) != (rhs.m_is_null == nullptr)) {
        return false;
    } else if (...